nexus/client.go

```diff
@@ -0,0 +1,177 @@
+//     Copyright 2020 Aicura Nexus Client and/or its authors
+//
+//     This file is part of Aicura Nexus Client.
+//
+//     Aicura Nexus Client is free software: you can redistribute it and/or modify
+//     it under the terms of the GNU Lesser General Public License as published by
+//     the Free Software Foundation, either version 3 of the License, or
+//     (at your option) any later version.
+//
+//     Aicura Nexus Client is distributed in the hope that it will be useful,
+//     but WITHOUT ANY WARRANTY; without even the implied warranty of
+//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//     GNU Lesser General Public License for more details.
+//
+//     You should have received a copy of the GNU Lesser General Public License
+//     along with Aicura Nexus Client.  If not, see <https://www.gnu.org/licenses/>.
+
+package nexus
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"go.uber.org/zap"
+)
+
+const (
+	userAgent              = "acuri-nexus-client"
+	applicationJSONContent = "application/json"
+	defaultNewAPIVersion   = "beta"
+	apiPath                = "service/rest"
+)
+
+type service struct {
+	client *Client
+}
+
+// Client base structure for the Nexus Client
+type Client struct {
+	httpClient *http.Client
+	baseURL    *url.URL
+	username   string
+	password   string
+	apiVersion string
+	shared     service // same instance shared among all services
```

Comment:
I'm not sure I get this one. What are we sharing across all services? The client itself? Why not just share the client directly if so?

```suggestion
	shared     *Client // same instance shared among all services
```

And even then, why is this part of the client? If I'm reading this correctly, we're inserting a `Client` in the `Client` definition itself. What do we plan on achieving with this? Can you share some example of how this would be used?

My gut tells me the client is supposed to be a Singleton, but I'm not entirely sure.

Reply:
The `Client`is singleton being shared amongst all the service, same pattern used in here:
https://github.com/google/go-github/blob/master/github/github.go#L159

Reply:
Hm, I'm starting to get it. So here we initialize a client and then assign it to each service we define in the client struct itself (such as the UserService), which I see we're doing. Then every time we need to access a service, we do it via this client singleton, like `NewDefaultClient(baseUrl).UserService.ListAllUsers()`.

We use this `shared` pointer when assigning the client pointer of each service, like `c.UserService = (*UserService)(&c.shared)`. But if I'm not mistaken, this is the only place we use this `shared` variable. Couldn't we keep it as a variable from the `NewClient` function instead of having a variable in the `Client` struct itself? I don't have a strong opinion about this and I think either way would be fine, I'm just trying to make sure I understand how this is being used.

And finally, we don't have any mechanism in place which prevents the caller from accidentally creating another instance of the client, right? Or am I missing something? Again, I don't necessarily see this as a problem, just making sure I caught up and we're on the same page. :-)

Reply:
We can't control how many clients a client puts on the heap, unfortunately. :(

Reply:
> We can't control how many clients a client puts on the heap, unfortunately. :(

Oh well, it should be fine while things are simple. In the future we can consider keeping an unexported `client` variable of type `*Client` and then check if this variable isn't `nil` when `NewClient` is called. We may create a new client at this point (and initialize this unexported variable), return the existing one or throw an error if the requested configuration doesn't match the existing client.

If we're that concerned with heap allocation, that is. Maybe there's even a valid use case for more than one client we don't see now. But enough of me digressing, I'm happy with how this is now :grin: 

> About having the shared in variable instead, I'd think won't matter at this point. I'd rather stick this way since seems to be a pattern and was proven good by others APIs. Could lead to problems that we are not seeing right now. :(

Yup, that makes sense.